Step5 : 로또(수동) 리뷰 요청드립니다.

Heonseop opened the pull request:
한 주간 친절하게 리뷰도 해주시고, 답변도 해주셔서 감사합니다.
이번으로 끝나는게 아니라 다음에 또 좋은 일로 만날 수 있으면 좋겠습니다.
이번 단계도 리뷰 잘 부탁드립니다.
감사합니다.

Heonseop committed:
feat : step4 를 위한 step3 코드 패키지 이동

Heonseop committed:
doc : step4 기능 요구사항 정리

Heonseop committed:
fix : step 3 추가 리뷰사항 수정

-StringUtil 생성자 수정
-Money divide 함수명 수정
-LottoBuyCount oneGameMoney 선언 위치 변경

Heonseop committed:
fix : LottoReward를 Rank로 변경
- 요구사항에 맞춰 Rank로 수정

Heonseop committed:
feat : 보너스 번호 입력 view 구현

Heonseop committed:
refactor : Game 클래스 1차 분리 및 잘못된 import 수정
- winLottoResult, lottoWinningStatistics 분리
- 잘못된 Test import 수정

Heonseop committed:
feat : BonusNumber 포함 확인을 위한 isContains 함수 구현

Heonseop committed:
feat : 당첨 통계 보너스볼 비교 추가 및 관련 outView 추가

Heonseop committed:
fix : getProfitPercent를 game 클래스에서 제거

Heonseop committed:
fix : 제한자 선언 순서 위배 수정

Heonseop committed:
fix : 주생성자와 부생성자 분리

Heonseop committed:
fix : 불필요한 결합을 낮추기 위해 Message를 사용하는 클래스 안으로 이동

Heonseop committed:
refactor : Rank valueOf BiPredicate 사용으로 변경

Heonseop committed:
refactor : Game class 함수 변경

- startLottoGame, getLottoResults 함수 통합

Heonseop committed:
refactor : 변수의 스코프 최소화 및 코드간 라인 분리

Heonseop committed:
refactor : LottoGenerator 초기화 함수 리팩토링
- IntStream으로 로또 객체 초기화 선언
- 클래스 로딩 때, 한번만 실행될 수 있도록 static 선언

Heonseop committed:
fix : lottoGenerator intStream rangeClosed 로 변경

Heonseop committed:
refactor : LottoResult 클래스 명을 Lotto로 변경

Heonseop committed:
doc : 5단계 기능 요구사항 정리

Heonseop committed:
refactor : WinningLotto 클래스 구현

Heonseop committed:
refactor : LottoNumber 정적 팩토리 적용 및 캐시 적용

Heonseop committed:
refactor : Lottos wrapping 클래스 구현

Heonseop committed:
feat : 수동으로 구매할 로또 수, 수동 로또 번호 입력, 통계 구현

Heonseop committed:
refactor : lottoGenerator auto와 manual 분리

Heonseop committed:
refactor : final 선언 가능한 케이스 수정

Heonseop committed:
fix : resolve conflict

devyonghee commented on README.md:
```diff
@@ -1,3 +1,11 @@
+# 5 단계: 로또(수동)
+
+## 추가 기능 요구 사항
+
+- 수동으로 구매할 로또 수 입력 구현
+- 수동으로 구매할 번호 입력 구현
+- 구입 금액에서 수동으로 구매한 금액 차액만큼 자동으로 로또 사는 기능 구현
```
요구사항 정리 잘해주신 것 같아요!! 👍👍👍

Heonseop replied:
감사합니다 ㅎㅎ

devyonghee commented on src/main/java/step4/model/LottoNumber.java:
```diff
@@ -4,24 +4,39 @@
-        int lottoNumber = convertNumber(text);
-        checkOutOfSize(lottoNumber);
-        this.number = lottoNumber;
+    static {
+        IntStream.rangeClosed(LottoConstant.LOTTO_MIN_NUM, LottoConstant.LOTTO_MAX_NUM)
+                .forEach(num ->
+                        lottoNumbers.put(num, new LottoNumber(num))
+                );
```
미리 메모리에 추가해서 재활용하는 것이 좋은 것 같아요!!
혹시 알고 계실지도 모르겠지만 플라이웨이트 패턴이라고 추가적으로 참고해보시면 좋을 것 같아 자료 제공만 해드리겠습니다!!
[Flyweight 패턴](https://velog.io/@hoit_98/%EB%94%94%EC%9E%90%EC%9D%B8-%ED%8C%A8%ED%84%B4-Flyweight-%ED%8C%A8%ED%84%B4)

Heonseop replied:
감사합니다 ㅎㅎ 이 부분은 다음 과제때 한번 구현해보겠습니다 ㅎㅎ

devyonghee commented on src/main/java/step4/model/Lottos.java:
```diff
@@ -0,0 +1,56 @@
+        return builder.toString();
+    }
+
+    public int matchCountAboutRank(WinningLotto winningLotto, Rank rank) {
+        return (int) lottos.stream().filter(lotto -> winningLotto.match(lotto).equals(rank)).count();
+    }
```
원하는 랭크로 필터링을 해서 계속 찾기 때문에 같은 로직이 여러번 반복될 것 같은데요 🤔
이 부분은 결과만 계산 할수 있도록 `Ranks` 로 반환해주는 역할까지만 부여해주는 건 어떠신가요??

Heonseop replied:
혹시 좀 더 comment 주실수 있으실지 문의드립니다. Ranks를 생성해서 반환해주려니 그다음을 어떻게 설계해야될지 생각이 않나네요 ㅠ

devyonghee replied:
여기에서 `Ranks` 라는 타입으로 반환해준다면
클라이언트 입장에서  `Ranks` 를 활용하여 원하는 원하는 `Rank`의 개수를 요청하는 방법도 좋을 것 같은데 어떠신가요? 🤔
`Ranks` 가 직접 어떤 `Rank` 가 몇개 존재하는지 개수를 반환해주는 구조를 생각해봤습니다.

devyonghee commented on src/main/java/step4/model/Lottos.java:
```diff
@@ -0,0 +1,56 @@
+        this.lottos = lottos;
+    }
+
+    public Lottos plus(Lottos otherLottos) {
+        return new Lottos(Stream.concat(lottos.stream(), otherLottos.lottos.stream()).collect(Collectors.toList()));
+    }
```
코드를 살펴보다보면 많은 부분에서 `stream` 동작이 한줄로 이어지는 것 같아요 🤔
혹시 엔터로 동작을 구분해줘서 가독성을 개선해주는 방향은 어떻게 생각하시나요??
제가 참고한 자료 첨부해드리겠습니다. [형식 맞추기 가로 정렬](https://github.com/Yooii-Studios/Clean-Code/blob/master/Chapter%2005%20-%20%ED%98%95%EC%8B%9D%20%EB%A7%9E%EC%B6%94%EA%B8%B0.md#%EA%B0%80%EB%A1%9C-%EC%A0%95%EB%A0%AC)

Heonseop replied:
흠 가독성 관련으로 고민이 많네요 ㅠㅠ 감사합니다.

devyonghee commented on src/main/java/step4/model/LottoBuyCount.java:
```diff
@@ -24,10 +24,36 @@ public LottoBuyCount(int count) {
+    }
+
+    private static int convertNumber(String text) {
+        int result;
+        try {
+            result = Integer.parseInt(text);
+        } catch (NumberFormatException e) {
+            throw new LottoFormatException(ErrorMessageConstant.NOT_NUMBER);
+        }
+        return result;
```
지금은 원시타입이기 때문에 해당되지 않겠지만 
가능하다면 초기화가 안된 로컬변수는 NPE 가 발생될 수 있기 때문에 조심하는게 좋다고 생각하는데
혹시 이부분에 대해서 어떻게 생각하시나요?? 🤔
```suggestion
        try {
            return Integer.parseInt(text);
        } catch (NumberFormatException e) {
            throw new LottoFormatException(ErrorMessageConstant.NOT_NUMBER);
        }
```

Heonseop replied:
말씀해주신대로 다른 곳에서 이렇게 할 경우 NPE 발생가능성이 있겠네요 null 체크 로직을 추가하겠습니다.

devyonghee commented on src/main/java/step4/model/LottoWinningStatistics.java:
```diff
@@ -1,44 +1,47 @@
+    }
+
+    private static Map<Rank, Integer> initLottoWinningStatistics() {
+        Map<Rank, Integer> lottoWinningStatistics = new LinkedHashMap<>();
         List<Rank> ranks = Arrays.asList(Rank.values());
         Collections.reverse(ranks);
         for (Rank rank : ranks) {
-            this.lottoWinningStatistics.put(rank, 0);
+            lottoWinningStatistics.put(rank, 0);
         }
+        return lottoWinningStatistics;
     }
```
함수형을 이용하면 다음과 같이도 작성할 수 있을 것 같아요!! 😆
참고만 해주세요!!🙇‍♂️🙇‍♂️🙇‍♂️

```java
Arrays.stream(Rank.values())
                .sorted(Comparator.reverseOrder())
                .collect(Collectors.toMap(rank -> rank, rank -> 0, (r1, r2) -> r1, LinkedHashMap::new))
```

Heonseop replied:
stream 만 잘 사용하면 코드를 많이 줄일수 있겠군요 감사합니다 좀 더 공부해봐야겠네요

devyonghee commented on src/main/java/step4/model/generator/LottoAutoGenerator.java:
```diff
@@ -0,0 +1,50 @@
+    private static final List<LottoNumber> lottoCandidateNumbers;
+    private final LottoBuyCount lottoBuyCount;
+
+    static {
+        lottoCandidateNumbers = IntStream.rangeClosed(LottoConstant.LOTTO_MIN_NUM, LottoConstant.LOTTO_MAX_NUM)
+                .mapToObj(LottoNumber::of).collect(Collectors.toList());
+    }
```
`LottoNumber` 에 있는 로직과 반복되는 것 같은데요 
코드의 응집도를 위해 `LottoNumber` 에게 번호들의 리스트를 요청하는 방향은 어떻게 생각하시나요?? 🤔

Heonseop replied:
오 여기가 중복되고 있는지는 생각 못했었네요 한번 수정해보겠습니다 ㅎㅎ

devyonghee commented on src/main/java/step4/model/Game.java:
```diff
@@ -1,37 +1,39 @@
-            result.add(lottoGenerator.createLottoResult());
-            index.plus();
+    public LottoBuyCount getLottoAutoBuyCount(LottoBuyCount manualLottoBuyCount) {
+        if (totalLottoBuyCount.isLessThan(manualLottoBuyCount)) {
+            throw new LottoFormatException(ErrorMessageConstant.MANUAL_BUY_LOTTO_GREATER_THAN_TOTAL_BUY_LOTTO);
         }
-        return result;
+        return totalLottoBuyCount.minus(manualLottoBuyCount);
```
이 부분은 `Game` 에서 마치 `auto` 가 들어올지 `manual`이 들어올지 미리 예상하고 count 계산을 해주고 있는 것 같아요.
논리적인 의존성을 가지고 있는 것 같은데 이부분에 대해서 어떻게 생각하시나요?? 🤔
제가 참고한 자료 남겨드리도록 하겠습니다!!  [G22 논리적 의존성은 물리적으로 드러내라](https://velog.io/@hellojihyoung/%EA%B0%9C%EB%B0%9C-%EB%8F%84%EC%84%9C-Clean-Code-17%EC%9E%A5-%EB%83%84%EC%83%88%EC%99%80-%ED%9C%B4%EB%A6%AC%EC%8A%A4%ED%8B%B1#-g22-%EB%85%BC%EB%A6%AC%EC%A0%81-%EC%9D%98%EC%A1%B4%EC%84%B1%EC%9D%80-%EB%AC%BC%EB%A6%AC%EC%A0%81%EC%9C%BC%EB%A1%9C-%EB%93%9C%EB%9F%AC%EB%82%B4%EB%9D%BC)

Heonseop replied:
한번 수정해보겠습니다.

devyonghee commented on src/main/java/step4/model/LottoNumber.java:
```diff
@@ -4,24 +4,39 @@
+        return lottoNumber;
+    }
+
+    public static LottoNumber of(String text) {
+        return of(convertNumber(text));
+    }
```
중요한 부분은 아니긴 하지만 참고하시면 좋을 것 같아 코멘트 남겨드립니다!

알고 계시겠지만 이펙티브 자바에서 정적 팩토리 메서드 사용을 권장하고 있는데요.
이 책에서는 명명 규칙을 소개하고 있는데
인자가 하나인 경우 `from`, 여러 인자로 만드는 경우에는 `of` 를 얘기하고 있는데요 
`valueOf` 를 활용하셔도 괜찮을 것 같은데 어떠신가요??

제가 참고한 링크 남겨드리도록 하겠습니다. [생성자 대신 정적 팩토리 메서드를 고려하라](https://sun-22.tistory.com/84)

Heonseop replied:
명명규칙을 확인해 봤는데 valueOf가 맞겠군요 ㅎㅎ
감사합니다 수정하겠습니다.

devyonghee commented on src/main/java/step4/model/LottoWinningStatistics.java:
```diff
@@ -1,44 +1,47 @@
 
-    private void initLottoWinningStatistics() {
+    public LottoWinningStatistics(Lottos lottos, WinningLotto winLotto) {
+        setLottoWinningStatistics(lottos, winLotto);
+        setTotalProfit();
```
코드의 응집도와 쉽게 파악하기 위해 개인적으로 순수 함수를 많이 이용하고 있는데요. 🤔
메소드 내부에서 필드 값을 필드가 언제 어떻게 변경되는지 쉽게 파악하기 어려울 것 같아요.
이 부분에 대해서 어떻게 생각하시나요??
알고 계실 수 있지만 제가 이용하고 있는 순수 함수에 대해서 자료 남겨드리겠습니다!!🙇‍♂️ [순수 함수](https://jeong-pro.tistory.com/23)

Heonseop replied:
혹시 좀 더 comment 주실수 있으실지 문의드립니다. 순수함수로 작성해보려니 어떻게 설계해야될지 생각이 않나네요 ㅠ

devyonghee replied:
메서드 내부에서는 외부의 값인 필드를 건들지 않고 
함수에서는 관련 인자를 받아 `totalProfit` 결과를 반환해주고 생성자에서는 그 결과를 할당해주는 구조입니다.

해당 메서드는 언제나 같은 동작을 하게 되며,
외부의 값도 수정하지 않기 때문에 재사용성이 높고 동작 예측이 가능한 장점이 있다고 생각됩니다. 🤔

반대로 메서드 내부에서 필드 할당을 숨기게 되면 어디서 필드가 변경되는지 찾기 어려워 유지보수가 힘들어진다고 생각돼서 드린 제안이었습니다 😆

devyonghee reviewed (changes requested):
안녕하세요 헌섭님!!
드디어 마지막 단계네요👍👍👍
헌섭님 말씀처럼 어디선가 좋은 인연으로 만나뵈면 좋을 것 같아요 😆😆

이번 5단계 로또 (수동) 도 잘 구현해주신 것 같아요!!
추가로 참고하시면 더 좋을 것 같은 
소소한 코멘트 몇가지 남겨드렸는데 확인 한번만 부탁드리겠습니다 🙇‍♂️

코멘트와 관련하여 궁금하거나 이상한 점이 보인다면 언제든 편하게 DM 이나 코멘트 부탁드리겠습니다. 🙏

Heonseop committed:
refactor : comment 수정

- LottoAutoGenerator 후보군 설정을 LottoNumber에서 조회하도록 수정
- LottoBuyCount NPE 체크
- Lottos plus 가독성 관련 수정
- initLottoWinningStatistics Stream으로 변경

Heonseop committed:
refactor : comment 2차 수정

- LottoNumber of를 규칙에 따라 ofValue로 수정
- convertNumber 함수 정리

Heonseop committed:
refactor : Game 클래스 논리적 책임 분리

devyonghee commented on src/main/java/step4/model/LottoWinningStatistics.java:
```diff
@@ -1,44 +1,44 @@
 
 public class LottoWinningStatistics {
-    private final Map<Rank, Integer> lottoWinningStatistics = new LinkedHashMap<>();
+    private static final Map<Rank, Integer> lottoWinningStatistics;
+
+    static {
+        lottoWinningStatistics = initLottoWinningStatistics();
+    }
```
불필요한 static 블럭은 사용하지 않으셔도 될 것 같아요!!😆

```suggestion
    private static final Map<Rank, Integer> lottoWinningStatistics = initLottoWinningStatistics();
```

devyonghee commented on src/main/java/step4/model/Lottos.java:
```diff
@@ -0,0 +1,64 @@
+        this.lottos = new ArrayList<>();
+    }
+
+    public Lottos(Lotto lotto) {
+        this.lottos = Stream.of(lotto).collect(Collectors.toList());
+    }
+
+    public Lottos(List<Lotto> lottos) {
+        this.lottos = lottos;
+    }
```
이전에 코멘트 해드려서 아시겠지만 부생서자가 주생성자를 이용하는 방식을 사용하면 코드 유지보수가 수월해질 것 같아요! 😁

devyonghee commented on src/main/java/step4/model/Lottos.java:
```diff
@@ -0,0 +1,64 @@
+        this.lottos = new ArrayList<>();
+    }
+
+    public Lottos(Lotto lotto) {
+        this.lottos = Stream.of(lotto).collect(Collectors.toList());
+    }
```
한 객체로 리스트를 생성하는 것이 목적이라면 싱글톤 리스트를 이용해보시면 좋을 것같아요!!
```suggestion
    public Lottos(Lotto lotto) {
        this.lottos = Collections.singletonList(lotto);
    }
```

devyonghee commented on src/main/java/step4/model/Money.java:
```diff
@@ -9,6 +9,10 @@ public class Money implements Comparable<Money> {
 
     private double money;
 
+    public Money() {
+        this.money = 0;
+    }
```
기본 생성자를 이용하는 방법도 좋지만 
`Money.zero()` 같은 정적 팩토리 메소드를 통해 의미를 명확하게 부여해주는 방향은 어떠신가요?🤔

devyonghee commented on src/main/java/step4/model/generator/LottoAutoGenerator.java:
```diff
@@ -0,0 +1,48 @@
+
+public class LottoAutoGenerator implements LottoGenerator {
+
+    private static final List<LottoNumber> lottoCandidateNumbers;
+
+    static {
+        lottoCandidateNumbers = LottoNumber.getLottoCandidateNumbers();
+    }
```
위에서 코멘트 해드려서 아시겠지만 
꼭 필요한 경우가 아니라면 직접 할당해도 좋을 것 같습니다!! 😆
```suggestion
    private static final List<LottoNumber> lottoCandidateNumbers = LottoNumber.getLottoCandidateNumbers();
```

devyonghee commented on src/main/java/step4/view/InputView.java:
```diff
@@ -24,4 +32,21 @@ public static String inputLottoBonusNumber() {
+            return result;
+        }
+        System.out.println(INPUT_BUY_MANUAL_LOTTOS);
+        for (LottoBuyCount index = new LottoBuyCount(0); index.isLessThan(lottoBuyCount); index.plus()) {
+            result.add(sc.nextLine());
+        }
```
뷰에서 도메인 객체를 생성해서 로직을 활용하는 방향이 괜찮은 방향인지 고민해보시면 좋을 것 같아요 🤔🤔🤔

devyonghee commented on src/test/java/step4/model/LottoManualGeneratorTest.java:
```diff
@@ -0,0 +1,32 @@
+    @DisplayName("로또생성 테스트")
+    @Test
+    void createLotto_pass_01() {
+        LottoManualGenerator lottoManualGenerator = new LottoManualGenerator(new ArrayList<String>() {{
+            add("1,2,3,4,5,6");
+        }}, new LottoBuyCount(1));
```
하나의 대상으로 리스트를 생성하는 것이라면
`Collections.singletonList` 를 활용하는 것이 더 적절할 것 같아요 🤔

devyonghee reviewed (approved):
헌섭님 마지막 단계까지 피드백 반영해주시느라 고생 많으셨습니다 🙇‍♂️
로또 미션을 진행하시면서 고민해주신 부분들이 많이 보이는 것 같아요 👍👍

많은 부분을 도와드리고자 피드백을 열심히 드리려고 했는데
리뷰어로써 부족하지는 않았는지 걱정이 되네요 🥲

추가적으로 한번 더 고민해보시면 좋을 것 같은 부분들 코멘트 남겨드렸습니다!
로또 미션 진행하시느라 정말 고생 많으셨고
남은 미션도 열심히 하셔서 꼭 좋은 결과 있으셨으면 좋겠습니다 😆👍
